src/pages/VenuePage/VenuePage.tsx

```diff
@@ -64,8 +73,62 @@ const VenuePage = () => {
     userPurchaseHistory: state.firestore.ordered.userPurchaseHistory,
     userPurchaseHistoryRequestStatus:
       state.firestore.status.requested.userPurchaseHistory,
+    retainAttendance: state.attendance.retainAttendance,
   }));
 
+  useEffect(() => {
+    const onClickWindow = (event: any) => {
+      // event.target.id &&
+      //   user &&
+      //   analytics.logEvent("clickonbutton", {
+      //     buttonId: event.target.id,
+      //     userId: user.uid,
+      //   });
+    };
+
+    const leaveRoomBeforeUnload = () => {
+      if (user && !retainAttendance) {
+        leaveRoom(user);
+      }
+    };
+    window.addEventListener("click", onClickWindow, false);
+    window.addEventListener("beforeunload", leaveRoomBeforeUnload, false);
+    return () => {
+      window.removeEventListener("click", onClickWindow, false);
+      window.removeEventListener("beforeunload", leaveRoomBeforeUnload, false);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  useEffect(() => {
+    const venueName = venue?.name ?? "";
+    const prevLocations = profile?.lastSeenIn ?? {};
+    const interval = setInterval(() => {
+      if (!user || !venueName || (venueName && prevLocations[venueName]))
+        return;
+      const updatedLastSeenIn = {
+        ...prevLocations,
+        [venueName]: currentTimeInUnixEpoch,
+      };
+      updateUserProfile(user.uid, {
+        lastSeenIn: updatedLastSeenIn,
+        lastSeenAt: new Date().getTime(),
+        room: venueName,
+      });
+    }, LOC_UPDATE_FREQ_MS);
+    return () => {
+      clearInterval(interval);
+      if (retainAttendance && user && profile) {
+        updateUserProfile(user.uid, {
+          lastSeenIn: { ...prevLocations, [venueName]: 0 },
+          lastSeenAt: new Date().getTime(),
+          room: null,
+        });
+      }
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
Guessing this is the cause of the location unloading? Would be great not to have this in the final PR.

Reply:
The problem here is that profile needs to be added in the dependency array and whenever this logic updates the location of the user, the profile is changed as well, and then this reruns and then it becomes an infinite loop

Reply:
If you have any suggestions or if statements that can avoid this feel free to post them here.

Reply:
Understood. This is hiding a bug and usually will prevent renders that should happen..

Reply:
What about changing updateUserProfile to call a function that only needs to receive the new attendance. Something where profile is no longer referenced inside the hook.

const { profile } = useUser()
const updateAttendance(user, loc, ts) { updateUserProfile(user.uid, {...profile, [loc]: ts}) }

Then you can remove the dependency on profile that’s inside the hook.